ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/PaginatingAsyncStream.scala

```diff
@@ -44,4 +45,41 @@ object PaginatingAsyncStream {
       }
       .flatMapConcat(Source(_))
   }
+
+  /**
+    * Concatenates the results of multiple asynchronous calls into
+    * a single [[Source]], passing the last seen event's ledger [[Offset]] and node index to the
+    * next iteration query, so it can continue reading events from this point.
+    *
+    * This is to implement pagination based on ledger offset and event node index.
+    * The main purpose of the pagination is to break down large queries
+    * into smaller batches. The reason for this is that we are currently using
+    * simple blocking JDBC APIs and a long-running stream would end up
+    * occupying a thread in the DB pool, severely limiting the ability
+    * of keeping multiple, concurrent, long-running streams while serving
+    * lookup calls.
+    *
+    * @param initialOffset initial ledger [[Offset]]
+    * @param extractOffsetAndNodeIndex function that extracts [[Offset]] and node index from the result entry of type [[T]]
+    * @param query a function that takes [[Offset]] and optional node index to start pagination from
+    * @tparam T the type of the items returned in each call
+    */
+  def streamFrom[T](initialOffset: Offset, extractOffsetAndNodeIndex: T => (Offset, Int))(
+      query: (Offset, Option[Int]) => Future[Vector[T]]
```

Comment:
```suggestion
  def streamFrom[Off, T, Ix](initialOffset: Off, extractOffsetAndNodeIndex: T => (Off, Ix))(
      query: (Off, Option[Ix]) => Future[Vector[T]]
```

Reply:
yeah, that makes sense, I will do this refactoring in a follow-up PR, don't want to wait for another build.

Reply:
@S11001001 is there any specific reason why you listed the type parameters in this order:
`[Off, T, Ix]`?

I would do `[Off, Ix, T]`.

Reply:
When in doubt I do first occurrence, left to right, just as an automatic generalization algorithm would likely choose. `Off, T, Ix` is the order of first occurrence. I don't care super hard as long as they're abstract.

Reply:
I was thinking about this as well, but initially thought that we might as well be very specific here. 
But of course, I have just noticed that we would need to do something similar for `CommandCompletionsReader`, which doesn't use any pagination at all at the moment, where we don't have a "node index" and would set the `Ix` type parameter to `Unit`.

Reply:
I would say that [a concrete type opposes the goal of specificity](https://typelevel.org/blog/2017/09/05/three-types-of-strings.html). It permits data mangling and conjuration from thin air, whereas the abstraction makes clear that the function is agnostic with respect to the data represented by the three tparams, and won't conjure any Offsets or Ints itself, never mind compare them or perform other operations.